trip_app/trips/tests/test_permissions.py

```diff
@@ -65,3 +65,40 @@ def test_admin_delete(admin_client):
     url = reverse("trips-detail", args=[trip.id])
     resp = admin_client.delete(url)
     assert resp.status_code == 204
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    "url", ["trip-requests-approve", "trip-requests-decline"]
+)
+def test_user_unable_approve_decline(client, create_user, url):
+    """ Verify that authenticated non driver, non admin can't approve(decline)
+        trip request with POST api/trip-requests/<id>/approve(decline). """
+    _, username, password = create_user
+    client.login(username=username, password=password)
+    trip_request = TripRequestFactory()
+    request_url = reverse(url, args=[trip_request.id])
+    resp = client.post(request_url)
+    assert resp.status_code == 403
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    "url, status",
+    [
+        ("trip-requests-approve", "Approved"),
+        ("trip-requests-decline", "Declined"),
+    ],
+)
+def test_owner_approve_decline(client, create_user, url, status):
+    """ Verify that driver can approve(decline) trip request with
+        POST api/trip-requests/<id>/approve(decline). """
+    _, username, password = create_user
+    user = User.objects.get(username=username)
+    client.login(username=username, password=password)
+    trip = TripFactory(driver=user)
+    trip_request = TripRequestFactory(trip=trip)
```

Comment:
candidates to be fixtures, better if globally reusable: `user(create_user)`, `trip(user)`, `trip_request(trip)`

Reply:
That's a good point. I used to abuse with such things, hopefully now this should be updated and all start up logic of tests should be replaced into fixtures where it makes sense. Thanks.